File: extension/src/mergeReview.ts
/**
 * Jessie Merge Review — Azure branch diff + Claude impact (web parity).
 */

import * as fs from 'fs';
import * as os from 'os';
import * as path from 'path';
import * as vscode from 'vscode';
import {
  ensureClaudeApiKey,
  getBackendUrl,
  getUserId,
  pickBranch,
  promptAzureConnection,
  workspaceIdFrom,
} from './azure';
import { streamSse } from './sse';

interface ImpactItem {
  title?: string;
  detail?: string;
  severity?: string;
  files?: string[];
  why?: string;
  how_to_verify?: string;
}

interface MergeComplete {
  type: 'complete';
  verdict: string;
  overall_score: number;
  grade: string;
  total_issues: number;
  critical_count: number;
  high_count: number;
  medium_count: number;
  low_count: number;
  missing_count: number;
  report_path: string;
  files_changed: number;
  lines_added: number;
  lines_removed: number;
  commits_count: number;
  impact_analysis?: {
    summary?: string;
    ui_changes?: ImpactItem[];
    functionality_changes?: ImpactItem[];
    expected_issues?: ImpactItem[];
    test_checklist?: string[];
    recommendation?: string;
    model?: string;
  };
  metadata?: Record<string, unknown>;
}

function writeImpactMarkdown(result: MergeComplete): string {
  const impact = result.impact_analysis || {};
  const meta = result.metadata || {};
  const lines: string[] = [
    '# Jessie — Claude Impact Report',
    '',
    `**Verdict:** ${result.verdict}`,
    `**Score:** ${result.overall_score}/100 (${result.grade})`,
    '',
    'Verdict = merge recommendation (approve / needs changes).',
    'Score = merge safety 0–100 (higher is safer).',
    '',
    `Branches: ${meta.head_branch ?? '?'} → ${meta.base_branch ?? '?'}`,
    `Files: ${result.files_changed}  ·  +${result.lines_added} / -${result.lines_removed}`,
    '',
    '## Summary',
    impact.summary || 'No Claude summary available.',
    '',
    '## UI changes users will notice',
  ];

  for (const item of impact.ui_changes || []) {
    lines.push(`### ● ${(item.severity || 'medium').toUpperCase()} — ${item.title}`);
    lines.push(item.detail || '');
    if (item.files?.length) lines.push(`Files: ${item.files.join(', ')}`);
    lines.push('');
  }

  lines.push('## Functionality / behaviour changes');
  for (const item of impact.functionality_changes || []) {
    lines.push(`### ● ${(item.severity || 'medium').toUpperCase()} — ${item.title}`);
    lines.push(item.detail || '');
    if (item.files?.length) lines.push(`Files: ${item.files.join(', ')}`);
    lines.push('');
  }

  lines.push('## Issues you may face');
  for (const item of impact.expected_issues || []) {
    lines.push(`### ● ${(item.severity || 'medium').toUpperCase()} — ${item.title}`);
    lines.push(item.detail || '');
    if (item.why) lines.push(`Why: ${item.why}`);
    if (item.how_to_verify) lines.push(`Verify: ${item.how_to_verify}`);
    if (item.files?.length) lines.push(`Files: ${item.files.join(', ')}`);
    lines.push('');
  }

  lines.push('## Test checklist');
  for (const c of impact.test_checklist || []) lines.push(`- [ ] ${c}`);

  const out = path.join(os.tmpdir(), `jessie-claude-impact-${Date.now()}.md`);
  fs.writeFileSync(out, lines.join('\n'), 'utf8');
  return out;
}

export class JessieMergeReview {
  private _statusBar: vscode.StatusBarItem;
  private _context: vscode.ExtensionContext;
  private _isRunning = false;

  constructor(statusBar: vscode.StatusBarItem, context: vscode.ExtensionContext) {
    this._statusBar = statusBar;
    this._context = context;
  }

  async start(): Promise<void> {
    if (this._isRunning) {
      vscode.window.showWarningMessage('A Jessie merge review is already running.');
      return;
    }

    const claudeKey = await ensureClaudeApiKey(this._context);
    if (!claudeKey) return;

    const conn = await promptAzureConnection(this._context);
    if (!conn) return;

    const preferredBase =
      conn.branches.find(b => b === 'main' || b === 'master' || b === 'develop') ??
      conn.branches[0];
    const baseBranch = await pickBranch(conn.branches, 'Select BASE branch (target)', preferredBase);
    if (!baseBranch) return;

    const preferredHead =
      conn.branches.find(b => b !== baseBranch) ?? conn.branches[0];
    const headBranch = await pickBranch(conn.branches, 'Select HEAD branch (source / feature)', preferredHead);
    if (!headBranch) return;

    if (baseBranch === headBranch) {
      vscode.window.showWarningMessage('Base and head branches must be different.');
      return;
    }

    this._isRunning = true;
    this._statusBar.text = '$(loading~spin) Jessie — merge review...';

    const backendUrl = getBackendUrl();
    const body = {
      platform: 'azure',
      user_id: getUserId(),
      workspace_id: workspaceIdFrom(`${conn.ref.org}/${conn.ref.project}/${conn.ref.repo}`),
      repo: conn.ref.repo,
      token: conn.token,
      azure_org: conn.ref.org,
      azure_project: conn.ref.project,
      mode: 'branch',
      base_branch: baseBranch,
      head_branch: headBranch,
      post_comments: false,
      triggered_by: 'extension',
      claude_api_key: claudeKey,
    };

    try {
      const result = await streamSse(`${backendUrl}/merge/review`, body, (event) => {
        if (event.type === 'progress') {
          this._statusBar.text = `$(loading~spin) Jessie — ${event.message}`;
        }
      });

      if (result.type !== 'complete') {
        const msg = String((result as any).message || 'Merge review failed');
        this._statusBar.text = '$(error) Jessie — merge review failed';
        vscode.window.showErrorMessage(msg);
        return;
      }

      const r = result as unknown as MergeComplete;
      this._statusBar.text = `$(sparkle) Jessie — ${r.verdict} (${r.overall_score})`;

      const summary =
        `Merge review: ${r.verdict} · score ${r.overall_score}/100 (${r.grade}) · ` +
        `${r.files_changed} files · ${r.total_issues} issues`;

      const choice = await vscode.window.showInformationMessage(
        summary,
        'Open Report',
        'Open Impact',
        'Dismiss',
      );

      if (choice === 'Open Report' && r.report_path) {
        await vscode.commands.executeCommand('vscode.open', vscode.Uri.file(r.report_path));
      } else if (choice === 'Open Impact') {
        const impactPath = writeImpactMarkdown(r);
        await vscode.commands.executeCommand('vscode.open', vscode.Uri.file(impactPath));
      }
    } catch (err: any) {
      this._statusBar.text = '$(error) Jessie — merge review failed';
      vscode.window.showErrorMessage(`Merge review error: ${err.message ?? 'Unknown error'}`);
    } finally {
      this._isRunning = false;
    }
  }
}

File: backend/api/merge_review.py
import asyncio
import json
import logging
import sqlite3
import time
from datetime import date
from pathlib import Path

from fastapi import APIRouter, HTTPException, Query, Request
from fastapi.responses import StreamingResponse
from pydantic import BaseModel

from agents.merge_reviewer.node import MergeReviewAgent
from gateway.auth_headers import auth_from_request, extract_auth_from_headers, get_team_id
from memory.store import DB_PATH

logger = logging.getLogger(__name__)
merge_router = APIRouter()


def _ensure_merge_table() -> None:
    DB_PATH.parent.mkdir(parents=True, exist_ok=True)
    with sqlite3.connect(DB_PATH) as conn:
        conn.execute(
            """
            CREATE TABLE IF NOT EXISTS merge_review_history (
                id             INTEGER PRIMARY KEY AUTOINCREMENT,
                workspace_id   TEXT,
                user_id        TEXT,
                date           TEXT,
                verdict        TEXT,
                overall_score  INTEGER,
                total_issues   INTEGER,
                critical_count INTEGER,
                report_path    TEXT,
                tokens_used    INTEGER,
                cost_estimate  REAL,
                created_at     TIMESTAMP DEFAULT CURRENT_TIMESTAMP
            )
            """
        )
        conn.commit()


_ensure_merge_table()


class MergeReviewRequest(BaseModel):
    platform: str
    user_id: str
    workspace_id: str
    repo: str = ""
    token: str = ""
    azure_org: str = ""
    azure_project: str = ""
    gitlab_project_id: str = ""
    mode: str = "branch"
    pr_number: int | None = None
    base_branch: str = "main"
    head_branch: str = ""
    from_sha: str = ""
    to_sha: str = ""
    post_comments: bool = False
    repo_path: str = ""
    triggered_by: str = "web"
    # Anthropic Claude API key (required — each user supplies their own)
    claude_api_key: str = ""


def _save_merge_history(*, workspace_id: str, user_id: str, result: dict, report_path: str, tokens_used: int = 0, cost: float = 0.0) -> int:
    with sqlite3.connect(DB_PATH) as conn:
        cur = conn.execute(
            """
            INSERT INTO merge_review_history
                (workspace_id, user_id, date, verdict, overall_score, total_issues,
                 critical_count, report_path, tokens_used, cost_estimate)
            VALUES (?, ?, ?, ?, ?, ?, ?, ?, ?, ?)
            """,
            (
                workspace_id,
                user_id,
                date.today().isoformat(),
                result.get("verdict", "needs_changes"),
                result.get("overall_score", 0),
                result.get("total_issues", 0),
                result.get("critical_count", 0),
                report_path,
                tokens_used,
                cost,
            ),
        )
        conn.commit()
        return int(cur.lastrowid)


def _write_merge_markdown(req: MergeReviewRequest, result: dict) -> str:
    reviews_dir = Path.cwd().parent / "reviews" / "merge"
    reviews_dir.mkdir(parents=True, exist_ok=True)
    report_name = f"merge-{req.workspace_id or 'workspace'}-{int(time.time())}.md"
    report_path = reviews_dir / report_name

    summary = result.get("change_summary") or {}
    impact = result.get("impact_analysis") or {}
    lines = [
        "# Jessie Merge Review",
        "",
        f"- Platform: {req.platform}",
        f"- Repository: {req.repo}",
        f"- Base branch: {result.get('metadata', {}).get('base_branch', req.base_branch)}",
        f"- Head branch: {result.get('metadata', {}).get('head_branch', req.head_branch)}",
        f"- Verdict: {result.get('verdict', 'needs_changes')}",
        f"- Overall score: {result.get('overall_score', 0)} ({result.get('grade', 'F')})",
        "",
        "## Claude Impact Analysis",
        impact.get("summary") or "No Claude summary available.",
        "",
        "### UI changes users will notice",
    ]
    for item in impact.get("ui_changes", []) or []:
        lines.append(f"- **{item.get('title', 'UI change')}** ({item.get('severity', 'medium')}): {item.get('detail', '')}")
        if item.get("files"):
            lines.append(f"  - Files: {', '.join(item.get('files') or [])}")
    lines.append("")
    lines.append("### Functionality / behaviour changes")
    for item in impact.get("functionality_changes", []) or []:
        lines.append(f"- **{item.get('title', 'Functionality change')}** ({item.get('severity', 'medium')}): {item.get('detail', '')}")
        if item.get("files"):
            lines.append(f"  - Files: {', '.join(item.get('files') or [])}")
    lines.append("")
    lines.append("### Issues you may face")
    for item in impact.get("expected_issues", []) or []:
        lines.append(f"- **{item.get('title', 'Issue')}** ({item.get('severity', 'medium')})")
        lines.append(f"  - What: {item.get('detail', '')}")
        lines.append(f"  - Why: {item.get('why', '')}")
        lines.append(f"  - Verify: {item.get('how_to_verify', '')}")
    lines.append("")
    lines.append("### Test checklist")
    for check in impact.get("test_checklist", []) or []:
        lines.append(f"- [ ] {check}")
    lines.append("")
    lines.append("## Risks")
    for issue in result.get("issues", []):
        lines.append(f"### [{issue.get('severity', 'low').upper()}] {issue.get('title', 'Issue')}")
        lines.append(issue.get("detail") or issue.get("description", ""))
        lines.append(f"Fix: {issue.get('fix') or issue.get('suggestion', '')}")
        if issue.get("code_snippet"):
            lines.append("```diff")
            lines.append(issue["code_snippet"])
            lines.append("```")
        lines.append("")

    lines.append("## Missing")
    for issue in result.get("missing_items", []):
        lines.append(f"### {issue.get('title', 'Missing')}")
        lines.append(issue.get("detail") or issue.get("description", ""))
        lines.append(f"Fix: {issue.get('fix') or issue.get('suggestion', '')}")
        if issue.get("code_snippet"):
            lines.append("```diff")
            lines.append(issue["code_snippet"])
            lines.append("```")
        lines.append("")

    lines.append("## Changed Files")
    for f in result.get("diff_files", []):
        lines.append(
            f"- `{f.get('filename', 'unknown')}` ({f.get('status', 'modified')}) "
            f"+{f.get('added', 0)} -{f.get('removed', 0)}"
        )

    report_path.write_text("\n".join(lines), encoding="utf-8")
    return str(report_path)


class AzureConnectRequest(BaseModel):
    platform: str
    repo: str
    token: str
    azure_org: str = ""
    azure_project: str = ""
    gitlab_project_id: str = ""


@merge_router.post("/open-prs")
async def open_prs(req: AzureConnectRequest):
    try:
        agent = MergeReviewAgent()
        prs = await asyncio.to_thread(
            agent.list_open_prs,
            platform=req.platform,
            repo=req.repo,
            token=req.token,
            azure_org=req.azure_org,
            azure_project=req.azure_project,
        )
        return {"prs": prs}
    except Exception as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


@merge_router.post("/branches")
async def list_branches(req: AzureConnectRequest):
    try:
        agent = MergeReviewAgent()
        branches = await asyncio.to_thread(
            agent.list_branches,
            platform=req.platform,
            repo=req.repo,
            token=req.token,
            azure_org=req.azure_org,
            azure_project=req.azure_project,
        )
        return {"branches": branches}
    except Exception as exc:
        raise HTTPException(status_code=400, detail=str(exc)) from exc


# Keep GET for compatibility, but prefer POST (token not in URL).
@merge_router.get("/open-prs")
async def open_prs_get(
    platform: str = Query(...),
    repo: str = Query(...),
    token: str = Query(...),
    azure_org: str = Query(""),
    azure_project: str = Query(""),
):
    return await open_prs(AzureConnectRequest(
        platform=platform, repo=repo, token=token,
        azure_org=azure_org, azure_project=azure_project,
    ))


@merge_router.get("/branches")
async def list_branches_get(
    platform: str = Query(...),
    repo: str = Query(...),
    token: str = Query(...),
    azure_org: str = Query(""),
    azure_project: str = Query(""),
):
    return await list_branches(AzureConnectRequest(
        platform=platform, repo=repo, token=token,
        azure_org=azure_org, azure_project=azure_project,
    ))

@merge_router.post("/review")
async def review_merge(req: MergeReviewRequest, request: Request):
    auth = auth_from_request(request, require_key=False)
    api_key = (auth.api_key or req.claude_api_key or "").strip()
    if not api_key:
        extract_auth_from_headers(
            api_key="", provider=None, user_id=None, workspace_id=None, require_key=True,
        )
    user_id = auth.user_id if auth.user_id != "anon" else req.user_id
    workspace_id = auth.workspace_id if auth.workspace_id != "default" else req.workspace_id
    provider = auth.provider if auth.api_key else "anthropic"

    async def _generate():
        queue: asyncio.Queue = asyncio.Queue()

        def on_progress(event: dict):
            queue.put_nowait(event)

        async def _run():
            start = time.monotonic()
            try:
                if not api_key:
                    await queue.put({
                        "type": "error",
                        "code": "api_key_required",
                        "message": "Include your Claude API key in the X-Claude-API-Key header.",
                    })
                    return

                agent = MergeReviewAgent()
                result = await agent.review(
                    platform=req.platform,
                    repo=req.repo,
                    token=req.token,
                    mode=req.mode,
                    user_id=user_id,
                    workspace_id=workspace_id,
                    base_branch=req.base_branch,
                    head_branch=req.head_branch,
                    pr_number=req.pr_number,
                    azure_org=req.azure_org,
                    azure_project=req.azure_project,
                    gitlab_project_id=req.gitlab_project_id,
                    post_comments=req.post_comments,
                    on_progress=on_progress,
                    claude_api_key=api_key,
                    provider=provider,
                )
                report_path = _write_merge_markdown(req, result)
                review_id = _save_merge_history(
                    workspace_id=req.workspace_id,
                    user_id=req.user_id,
                    result=result,
                    report_path=report_path,
                )
                duration = time.monotonic() - start
                result["report_path"] = report_path
                result["duration_seconds"] = round(duration, 1)
                result["review_id"] = review_id
                await queue.put({"type": "complete", **result})
            except Exception as exc:
                logger.exception("Merge review failed")
                await queue.put({"type": "error", "code": "merge_review_failed", "message": str(exc)})
            finally:
                await queue.put(None)

        asyncio.create_task(_run())
        while True:
            event = await queue.get()
            if event is None:
                break
            yield f"data: {json.dumps(event)}\n\n"
        yield "data: [DONE]\n\n"

    return StreamingResponse(
        _generate(),
        media_type="text/event-stream",
        headers={"Cache-Control": "no-cache", "X-Accel-Buffering": "no"},
    )


@merge_router.get("/history/{workspace_id}")
async def merge_history(workspace_id: str):
    with sqlite3.connect(DB_PATH) as conn:
        rows = conn.execute(
            """
            SELECT date, verdict, overall_score, total_issues, critical_count, report_path,
                   cost_estimate, created_at
            FROM merge_review_history
            WHERE workspace_id = ?
            ORDER BY created_at DESC
            LIMIT 20
            """,
            (workspace_id,),
        ).fetchall()
    return [
        {
            "date": r[0],
            "verdict": r[1],
            "overall_score": r[2],
            "total_issues": r[3],
            "critical_count": r[4],
            "report_path": r[5],
            "cost_estimate": r[6],
            "created_at": r[7],
        }
        for r in rows
    ]
